### PR TITLE
Add informer lag histograms

### PR DESCRIPTION
**Pull request by adleong**
In order to detect if the destination controller's k8s informers have fallen behind, we add a histogram for each resource type.  These histograms track the delta between when an update to a resource occurs and when the destination controller processes that update.  We do this by looking at the timestamps on the managed fields of the resource and looking for the most recent update and comparing that to the current time.

The histogram metrics are of the form `{kind}_informer_lag_ms_bucket_*`.

* We record a value only for updates, not for adds or deletes.  This is because when the controller starts up, it will populate its cache with an add for each resource in the cluster and the delta between the last updated time of that resource and the current time may be large.  This does not represent informer lag and should not be counted as such.
* When the informer performs resyncs, we get updates where the updated time of the old version is equal to the updated time of the new version.  This does not represent an actual update of the resource itself and so we do not record a value.
* Since we are comparing timestamps set on the manged fields of resources to the current time from the destination controller's system clock, the accuracy of these metrics depends on clock drift being minimal across the cluster.
* We use histogram buckets which range from 500ms to about 17 minutes.  In my testing, an informer lag of 500ms-1000ms is typical.  However, we wish to have enough buckets to identify cases where the informer is lagged significantly behind.

### CODE REVIEW

File: controller/api/destination/watcher/prometheus.go [hunk 1]
```diff
@@ -34,6 +34,68 @@ type (
...
+			Name:    "endpointslice_informer_lag_ms",
+			Help:    "The amount of time between when an EndpointSlice resources is updated and when an informer observes it",
+			Buckets: informer_lag_ms_buckets,
+		},
+	)
+
+	serviceInformerLag = promauto.NewHistogram(
+		prometheus.HistogramOpts{
+			Name:    "service_informer_lag_ms",
+			Help:    "The amount of time between when a Serivce resources is updated and when an informer observes it",
```
Comment (alpeb) [inline controller/api/destination/watcher/prometheus.go:69]:
```suggestion
			Help:    "The amount of time between when a Service resource is updated and when an informer observes it",
```

File: controller/api/destination/watcher/prometheus.go [hunk 2]
```diff
@@ -34,6 +34,68 @@ type (
...
+			Name:    "endpoints_informer_lag_ms",
+			Help:    "The amount of time between when an Endpoints resources is updated and when an informer observes it",
+			Buckets: informer_lag_ms_buckets,
+		},
+	)
+
+	endpointsliceInformerLag = promauto.NewHistogram(
+		prometheus.HistogramOpts{
+			Name:    "endpointslice_informer_lag_ms",
+			Help:    "The amount of time between when an EndpointSlice resources is updated and when an informer observes it",
```
Comment (alpeb) [inline controller/api/destination/watcher/prometheus.go:61]:
```suggestion
			Help:    "The amount of time between when an EndpointSlice resource is updated and when an informer observes it",
```

File: controller/api/destination/watcher/prometheus.go [hunk 3]
```diff
@@ -34,6 +34,68 @@ type (
...
+			Name:    "service_informer_lag_ms",
+			Help:    "The amount of time between when a Serivce resources is updated and when an informer observes it",
+			Buckets: informer_lag_ms_buckets,
+		},
+	)
+
+	serverInformerLag = promauto.NewHistogram(
+		prometheus.HistogramOpts{
+			Name:    "server_informer_lag_ms",
+			Help:    "The amount of time between when a Server resources is updated and when an informer observes it",
```
Comment (alpeb) [inline controller/api/destination/watcher/prometheus.go:77]:
```suggestion
			Help:    "The amount of time between when a Server resource is updated and when an informer observes it",
```

Comment (alpeb) [inline controller/api/destination/watcher/prometheus.go:77]:
```suggestion
			Help:    "The amount of time between when a Server resource is updated and when an informer observes it",
```

File: controller/api/destination/watcher/prometheus.go [hunk 4]
```diff
@@ -34,6 +34,68 @@ type (
...
+			Buckets: informer_lag_ms_buckets,
+		},
+	)
+
+	podInformerLag = promauto.NewHistogram(
+		prometheus.HistogramOpts{
+			Name:    "pod_informer_lag_ms",
+			Help:    "The amount of time between when a Pod resources is updated and when an informer observes it",
```
Comment (alpeb) [inline controller/api/destination/watcher/prometheus.go:85]:
```suggestion
			Help:    "The amount of time between when a Pod resource is updated and when an informer observes it",
```

File: controller/api/destination/watcher/prometheus.go [hunk 5]
```diff
@@ -34,6 +34,68 @@ type (
...
+			Buckets: informer_lag_ms_buckets,
+		},
+	)
+
+	serviceProfileInformerLag = promauto.NewHistogram(
+		prometheus.HistogramOpts{
+			Name:    "serviceprofile_informer_lag_ms",
+			Help:    "The amount of time between when a ServiceProfile resources is updated and when an informer observes it",
```
Comment (alpeb) [inline controller/api/destination/watcher/prometheus.go:93]:
```suggestion
			Help:    "The amount of time between when a ServiceProfile resource is updated and when an informer observes it",
```

File: controller/k8s/k8s.go [hunk 6]
```diff
@@ -11,7 +11,7 @@ import (
 	"k8s.io/client-go/tools/cache"
 )
 
-const ResyncTime = 10 * time.Minute
+const ResyncTime = 5 * time.Minute
```
Comment (alpeb) [inline controller/k8s/k8s.go:14]: Why this change?

File: controller/api/destination/watcher/prometheus.go [hunk 7]
```diff
@@ -34,6 +34,68 @@ type (
...
+		250000,  // 4m 10s
+		1000000, // 16m 40s
+	}
+	endpointsInformerLag = promauto.NewHistogram(
+		prometheus.HistogramOpts{
+			Name:    "endpoints_informer_lag_ms",
+			Help:    "The amount of time between when an Endpoints resources is updated and when an informer observes it",
+			Buckets: informer_lag_ms_buckets,
+		},
+	)
```
Comment (alpeb) [inline controller/api/destination/watcher/prometheus.go:56]: Can we somehow condition this declaration to whether endpointsslices is enabled or not? If it is, which is the default nowadays, I think these metrics will always show up empty.
... or we can leave this in and just wait till we're able to drop support for endpoints in favor of endpointslices altogether, which should happen pretty soon if not already.

**Review by alpeb — COMMENTED** [review]
This is so good! Tests perfectly :+1:

File: controller/k8s/k8s.go (hunk 6 above)
Comment (adleong) [reply controller/k8s/k8s.go:14]: unintended, good catch!

File: controller/api/destination/watcher/prometheus.go (hunk 7 above)
Comment (adleong) [reply controller/api/destination/watcher/prometheus.go]: I think it's simplest to just leave them in.  It's just one histrogram without any additional labels, so I don't think it should be a problem.

**Review by alpeb — APPROVED** [review]

**Comment by olix0r** [comment]
> The histogram metrics are of the form `{kind}_informer_lag_ms_bucket_*`.

We haven't been rigorous about this previously, but [prom best practices](https://prometheus.io/docs/practices/naming/) encourage use of _seconds_ (i.e. with fractional float values) for all time units.

File: controller/api/destination/watcher/prometheus.go [hunk 8]
```diff
@@ -34,6 +34,68 @@ type (
...
+		prometheus.HistogramOpts{
+			Name:    "endpoints_informer_lag_secs",
+			Help:    "The amount of time between when an Endpoints resource is updated and when an informer observes it",
+			Buckets: informer_lag_secs_buckets,
+		},
+	)
+
+	endpointsliceInformerLag = promauto.NewHistogram(
+		prometheus.HistogramOpts{
+			Name:    "endpointslice_informer_lag_secs",
```
Comment (alpeb) [inline controller/api/destination/watcher/prometheus.go:60]: I've just noticed, I think it's better to use the plural form to have it aligned with the other metric `endpointslices_cache_size`
```suggestion
endpointslices_informer_lag_secs
```

File: controller/api/destination/watcher/prometheus.go [hunk 9]
```diff
@@ -34,6 +34,68 @@ type (
...
+			Help:    "The amount of time between when an EndpointSlice resource is updated and when an informer observes it",
```
Comment (alpeb) [inline controller/api/destination/watcher/prometheus.go:61]:
```suggestion
			Help:    "The amount of time between when an EndpointSlices resource is updated and when an informer observes it",
```

File: controller/api/destination/watcher/prometheus.go [hunk 10]
```diff
@@ -34,6 +34,68 @@ type (
...
+		10,   // 10s
+		25,   // 25s
+		50,   // 50s
+		100,  // 1m 40s
+		250,  // 4m 10s
+		1000, // 16m 40s
+	}
+	endpointsInformerLag = promauto.NewHistogram(
+		prometheus.HistogramOpts{
+			Name:    "endpoints_informer_lag_secs",
```
Comment (olix0r) [inline controller/api/destination/watcher/prometheus.go:52]: The [convention](https://prometheus.io/docs/practices/naming/) is specifically to use _seconds_
```suggestion
			Name:    "endpoints_informer_lag_seconds",
```

File: controller/api/destination/watcher/prometheus.go (hunk 9 above)
Comment (olix0r) [reply controller/api/destination/watcher/prometheus.go:61]: I believe it is correct as it is:
```
:; k get endpointslices.discovery.k8s.io -o yaml |grep kind
  kind: EndpointSlice
kind: List
```

**Review by olix0r — APPROVED** [review]
